Various small features added to tzxcleanup, tzxwav and tzxls

oxidaan opened the pull request:
With your great tools I was able to recover 90% of my spectrum tapes. While doing so I added some features to tzxtools.
I am both new to python and new to github, (I have a C++ background) so forgive me if something is not according to the best practices...

tzxwav now shows the time (hours:min:secs) (from wave/tape start) of each found frame (when verbose).
tzxwav skips showing really short blocks (<10 bytes) even when verbose, to avoid some output.

tzxcleanup shows total number of bytes written, this gives a good indication of the quality of the original tzx (as created by tzxwav)
tzxcleanup can now skip data blocks that have no matching (zx spectrum) header, so only matching header/data pairs remain
Activated with commandline argument -H --headermustmatch
tzxcleanup shows number of headerless blocks skipped due to the above.
tzxcleanup fixed a rarely seen crash (index out of range). Had this with a old tzx file (not created with tzxwav). If needed I can try to roll back and reproduce?

tzxls with shartup argument -w /--wide can now show output comma separated on one line.

tzxlib a zx spectrum header is now allowed to be 20 bytes long as well (so 19 or 20) Had lots of headers with this length, especially games, that do work on real zx spectrum, but were otherwise showed as 'bogus'. Probably by copy program I used in the 1980's.

tzxlib (tzxls/tzxwav) show a 16 byte hex dump for each block when using verbose. Sometimes gives some a clue what it is (especially when bogus header)

shred commented:
First of all, thank you for your pull request and the time you spent improving tzxtools!

I had a look at your changes, and sadly I found some problems. Actually, I wish you had created separated pull requests for each change.

First of all, I like your change that `tzxwav` also shows the time of each block found. :+1:
I would also merge your changes to `tzxcleanup`. :+1:

Now for the "but" part that always spoils the good spirit.

- I think `tzxwav` should show all blocks that were found, even the short ones. It may be confusing that they are not shown in the verbose output, but are part of the tzx file later. In my experience those short blocks also occur very rarely, unless `tzxwav` finds a lot of broken blocks.
- Can you give a use case for the `--wide` option? I tried it out, but found the result just unreadable and cannot see any practical advantage over the classic output.
- If `--wide` is activated, the output ends with a `,`, but no line feed. The command prompt is then at the end of that long line.
- The new options need to be added to the documentation as well. :wink:
- To be honest, I don't really like the `--hexdump` option as it clutters the output. Usually `tzxcat` can be used to generate a hex dump of a block. The `hexdump` library that you used need to be added to the `setup.py` as `install_requires`, so it is automatically installed as dependency by pip. However, the `hexdump` library is unable to convert the ZX Spectrum charset. You should use something similar to `convert.py`'s `convertToDump()` function instead. But I would prefer not to add this option at all.
- `tzxwav` immediately fails for me because it cannot find `TapFile` (it's missing in the imports).
- Removing the `--hexdump` option would also remove the global `showHexSample` variable that causes a bit of trouble.
- 20 byte size headers are not specified, and thus should not be supported. I also suspect that the 20th byte is just a 00 byte that was added because tzxwav detected a surplus pulse at the end of the recording.

Again, thanks for the pull request, and sorry for all those negative points.

oxidaan commented:
First: sorry for me committing it al-at-once, must be my lack of gitgub experience!...
> I think tzxwav should show all blocks that were found
The skipping of short blocks was indeed a change that, when I reviewed later, I had some doubts on too... It was one of the first things I did.
>Can you give a use case for the --wide option?
I used it (combined with tzxcleanup) to create compact list of all working programs found on the tape. I then copied/pasted that to an excell sheet. The trailing comma, indeed, should be removed. Could not find a quick way, in Python, to check if it is in the last entry of the loop.
> As for the hex dump, my main reason to add it was because I had *lots* of bogus headers, and had no clue what they where. (Is it just another Manic Miner copy or my brillant code created by my 16 year old self??) With hexdump I could at least see what I was missing. Later these bugus headers turned out to be caused by that 19/20 byte header issue (see below) That said, I improved a bit on the hexdump - allign to a fixed column position and skipped the address at begining which is always 00000000. But indeed, it needs hexdump library, can imagine you don't want that.
> tzxwav immediately fails for me...
Sorry, will check too... 
> Removing the --hexdump option 
... actually I think I committed that after the pull request? Not sure...
> 20 byte size headers are not specified
Well almost all my old games had such headers, and these work at the real Spectrum. imho when the real spectrum can load it, so should the tzx tools. Tried on zxspin, can load it too. Will try some more emulators... 
(For some weird reason only my games have these issue, I actually think the copy program (Mr copy) I used it the 80's was to blame)

Note I also had a (rare) crash reading a tzx file created by other software, some index error, which I fixed, but will try to reproduce, will follow up on that....

oxidaan commented:
> tzxwav immediately fails for me...
Ok I fixed that just now. My execuse might be he cause of this problem was pushed to github after I created my pull request - was not aware that later pushes end-up in an earlier pull request (?)

oxidaan commented:
>I think tzxwav should show all blocks that were found...

I removed the skipping of short data (in the output) of tzxwav is at loader.py line 70. I added it because especially at the beginning of my tapes I have a lot of 1/2/3 byte data lines, all with crc errors. But indeed doesn't make sense since tzxls still also shows it.

oxidaan commented:
Fixed the last comma separator at tzxls -w, added extra newline at end. Note at Windows, this newline doesn't seem to be needed.

shred commented:
I'm sorry for the slow progress. I'm a bit occupied at the moment, but I will have a look at your changes.

shred commented:
Thank you for your patience!

About the `--wide` option: Pasting a file list into Excel is certainly a corner case. It will also break if a file name contains a comma, so sometimes the output isn't correct. It's easy to get the same result by piping the output to a post processor, e.g. `tzxls mytape.tzx | tr '\n' ','`. I'm sorry, but for this reason, I think I won't merge the wide option. (This is why I wished you had made separate merge requests for each new feature.)

I'm also not convinced about the `--hexdump` option yet. I would prefer not to use the external dump library here, but to use the internal hexdump converter, as it converts all ZX Spectrum special characters.

I still like your change in `tzxcleanup`, though. I also like that `tzxwav` shows the time of each block. I would merge these changes.

By the way, don't worry about doing many commits. I will squash them all to a single commit before merging. :smiley:

oxidaan commented:
Hi, 

I totally understand that you don't like to pull-in extra library's just
for the hexdump feature. I had doubts with it myself. I didn't like my
own solution passing the corresponding option as a global either... I
just found it usefull in the beginning when I had lots of bogus headers
caused by that weird extra byte it is reading at all my games. Just
wanted to know what the bogus headers acually were. ...In the future I
should think better what changes are handy just for me and which changes
are handy for the community before create a pull request!... 

That said about those bogus headers, when ignoring that extra byte I had
much better results. All those headers seem to be created with a program
'Mr Copy' I used in the 1980's.
(https://spectrumcomputing.co.uk/entry/8470/ZX-Spectrum/Mr_Copy) Not
sure how widely that was used... 

As for the wide option, imho it doesn't hurt either, and your suggestion
with piping through tr doesn't work at Windows... I found it handy to
create a compact list of all working software on one text line (I
combined it with my new 'header must match' option). 

Let me thank you again for your amazing set of tools! At the moment I am
finished with all my old tapes and probably have 95% of my old software
back. Now my wife wants me to throw away my tapes ;-(... so I am afraid
I won't use your tools much more... 

Thanks again! 

Daan. 

Richard Körber schreef op 2021-02-01 23:42:

> Thank you for your patience! 
> 
> About the --wide option: Pasting a file list into Excel is certainly a corner case. It will also break if a file name contains a comma, so sometimes the output isn't correct. It's easy to get the same result by piping the output to a post processor, e.g. tzxls mytape.tzx | tr '\n' ','. I'm sorry, but for this reason, I think I won't merge the wide option. (This is why I wished you had made separate merge requests for each new feature.) 
> 
> I'm also not convinced about the --hexdump option yet. I would prefer not to use the external dump library here, but to use the internal hexdump converter, as it converts all ZX Spectrum special characters. 
> 
> I still like your change in tzxcleanup, though. I also like that tzxwav shows the time of each block. I would merge these changes. 
> 
> By the way, don't worry about doing many commits. I will squash them all to a single commit before merging. 😃 
> 
> --
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub [1], or unsubscribe [2].

 

Links:
------
[1] https://github.com/shred/tzxtools/pull/8#issuecomment-771210773
[2]
https://github.com/notifications/unsubscribe-auth/AE36RDM2NWTNJYCLFFRFSLLS44U5LANCNFSM4WCETA7Q

shred commented:
Yes, the tapes consume much space, considering how many thousand tapes could be stored on a single USB stick. But I couldn't throw them away either. :wink:

Is it okay if I still use your `tzxwav` and `tzxcleanup` changes? I will mention you as the author of course. If you like, you can also create a new merge request just with these changes.

Thank you!

oxidaan commented:
Hi,Of course you can use my changes! Thank you,Gr, Daan.Verzonden vanaf mijn Galaxy
-------- Oorspronkelijk bericht --------Van: Richard Körber <notifications@github.com> Datum: 13-02-21  17:14  (GMT+01:00) Aan: shred/tzxtools <tzxtools@noreply.github.com> Cc: oxidaan <daan@scherft.nl>, Author <author@noreply.github.com> Onderwerp: Re: [shred/tzxtools] Various small features added to tzxcleanup, tzxwav and tzxls (#8) 

Yes, the tapes consume much space, considering how many thousand tapes could be stored on a single USB stick. But I couldn't throw them away either. 😉

Is it okay if I still use your tzxwav and tzxcleanup changes? I will mention you as the author of course. If you like, you can also create a new merge request just with these changes.

Thank you!



—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

shred commented:
I have incorporated your changes in commit c7f92e71 and c4f36094.
